File: src/python/challenges/problems/count_negative_numbers_in_a_sorted_matrix_test.py
# https://leetcode.com/problems/count-negative-numbers-in-a-sorted-matrix

from typing import List
from core.problem_base import *

class CountNegativeNumbersInASortedMatrix(ProblemBase):
    def Solution(self, grid: List[List[int]]) -> int:
        result = 0
        height = len(grid)
        width = len(grid[0])

        for row in range(height):
            for col in range(width - 1, -1, -1):
                if grid[row][col] < 0:
                    result += height - row
                    width = col
                else:
                    width = col + 1
                    break

        return result

if __name__ == '__main__':
    TestGen(CountNegativeNumbersInASortedMatrix) \
        .Add(lambda tc: tc.Param([[3,2],[-3,-3],[-3,-3],[-3,-3]]).Result(6)) \
        .Add(lambda tc: tc.Param([[4,3,2,-1],[3,2,1,-1],[1,1,-1,-2],[-1,-1,-2,-3]]).Result(8)) \
        .Add(lambda tc: tc.Param([[3,2],[1,0]]).Result(0)) \
        .Run()

File: src/go/challenges/problems/count_negative_numbers_in_a_sorted_matrix_test.go
// https://leetcode.com/problems/count-negative-numbers-in-a-sorted-matrix
package problems

import (
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type CountNegativeNumbersInASortedMatrix struct{}

func TestCountNegativeNumbersInASortedMatrix(t *testing.T) {
	gen := core.TestSuite[CountNegativeNumbersInASortedMatrix]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param([][]int{{4, 3, 2, -1}, {3, 2, 1, -1}, {1, 1, -1, -2}, {-1, -1, -2, -3}}).Result(8)
	}).Add(func(tc *core.TestCase) {
		tc.Param([][]int{{3, 2}, {1, 0}}).Result(0)
	}).Add(func(tc *core.TestCase) {
		tc.Param([][]int{{3, 2}, {-3, -3}, {-3, -3}, {-3, -3}}).Result(6)
	}).Run(t)
}

func (CountNegativeNumbersInASortedMatrix) Solution(grid [][]int) int {
	result := 0
	height := len(grid)
	width := len(grid[0])

	for row := 0; row < height; row++ {
		for col := width - 1; col >= 0; col-- {
			if grid[row][col] < 0 {
				result += height - row
				width = col
			} else {
				width = col + 1
				break
			}
		}
	}

	return result
}
